src/parser/core/EventFilter.js

```diff
@@ -0,0 +1,32 @@
+class EventFilter {
+  eventType;
+  constructor(eventType) {
+    this.eventType = eventType;
+  }
+  _by;
+  by(value) {
+    if (value === undefined) {
+      return this._by;
+    }
+    this._by = value;
```

Comment:
I feel like one or both of these should be in these setters:
1. Clear functionality definition when calling these setters more than once: is it supposed to overwrite the existing filter? That's not really standard builder convention -- maybe it should throw a state error if _by is already set and you call this again? Or should it instead set `this._by = this._by | value;` ?
2. Parameter checking: if one of these values expects one of the enum values, check to make sure it exists?

Also, did you consider having the EventFilter object actually have a filter() function in it that applies its filters to the given event, rather than doing the compiling in the emitter to prepend the checks? I feel like you would eliminate like 10-20 lines of code and it would make more design sense to just have it be inside the filter object.

Reply:
In my experience overwriting the existing filter is quite common, as well as making all these methods functional but that's not as performant and I can't imagine this will ever go wrong.

They're not enum values, they're flags. Hard to check for all possible combinations and again there's little to gain I think.

I hadn't considered it, having the filters inside EventFilter does sound appealing, but a big advantage of the compilation is that it's the most performant. With all this logic in the EventEmitter - which I don't think is a bad place for the filter checking either, making EventFilter a dumb struct - it's already well isolated so should be no issue.

Reply:
All of my flags are enums in C#, that's why I used that name 😄  (C# has a [Flags] attribute you can apply to enums) -- but what I mainly meant was that you'd just validate like this:

```javascript
const VALID_FLAGS = SELECTED_PLAYER | SELECTED_PLAYER_PET | OTHER_PLAYER /*etc*/;
validate(flags) {
  return flags & VALID_FLAGS == flags;
}
```
It is a small gain, but it's also pretty small cost IMO

Wouldn't it be more performant to not have to compile at all? Having the filtering done inside the event filter should perform exactly the same outside of just skipping the compiling steps. (instead of prepending after you've built the event, you add in steps to the filter() function as the .spell, .by, etc. functions are called; meaning if .to is never called you are never even performing the if(filter.to) check)

Reply:
I could be completely visualizing it wrong though. Could you push to test maybe?

Reply:
I havent tested this or anything and I don't know if this works like I think it would (really not a javascript guru), but this is my general idea below. 

```javascript
  by(value) {
    // Could validate here :]
    const requiresSelectedPlayer = value & SELECTED_PLAYER;
    const requiresSelectedPlayerPet = value & SELECTED_PLAYER_PET;

    let check;
    if (requiresSelectedPlayer && requiresSelectedPlayerPet) {
      check = event => this.owner.byPlayer(event) || this.owner.byPlayerPet(event);
    } else if (requiresSelectedPlayer) {
      check = event => this.owner.byPlayer(event);
    } else if (requiresSelectedPlayerPet) {
      check = event => this.owner.byPlayerPet(event);
    }

    // if we validated we could move this if up and exit early instead
    if(check) {
      const old = this.filter;
      this.filter = function(event) {
        if(!check(event)) {
          return false;
        }
        return old();
      };
    }
    
    return this;
  }

filter(event) {
  return true;
}

```

This would functionally add filters with later calls instead of overwrite them though -- to do it the other way you'd just save _by, _spell, and _to as functions and then filter(event) probably need to call all three (even if they just return true;), which is probably what you were taking into account when you said it would be a performance hit

Reply:
As I said, I kinda like that EventFilter acts like a simple struct, I'm not really convinced moving this logic would be an improvement. I did implement everything else though.